src/sql/workbench/api/node/extHostNotebook.ts

```diff
@@ -6,24 +6,25 @@
 
 import * as sqlops from 'sqlops';
 import * as vscode from 'vscode';
+
 import { TPromise } from 'vs/base/common/winjs.base';
 import { IMainContext } from 'vs/workbench/api/node/extHost.protocol';
 import { Disposable } from 'vs/workbench/api/node/extHostTypes';
 import { localize } from 'vs/nls';
-
+import URI, { UriComponents } from 'vs/base/common/uri';
 
 import { ExtHostNotebookShape, MainThreadNotebookShape, SqlMainContext } from 'sql/workbench/api/node/sqlExtHost.protocol';
-import URI, { UriComponents } from 'vs/base/common/uri';
-import { INotebookManagerDetails } from 'sql/workbench/api/common/sqlExtHostTypes';
+import { INotebookManagerDetails, INotebookSessionDetails, INotebookKernelDetails, INotebookFutureDetails, FutureMessageType } from 'sql/workbench/api/common/sqlExtHostTypes';
+
+type Adapter = sqlops.nb.NotebookProvider | sqlops.nb.NotebookManager | sqlops.nb.ISession | sqlops.nb.IKernel | sqlops.nb.IFuture;
 
 export class ExtHostNotebook implements ExtHostNotebookShape {
 	private static _handlePool: number = 0;
 
 	private readonly _proxy: MainThreadNotebookShape;
-	private _providers = new Map<number, sqlops.nb.NotebookProvider>();
+	private _adapters = new Map<number, Adapter>();
```

Comment:
Helpful info: I refactored Adapters so that all tracked objects are stored here. Whenever we send "something" to the main thread - be it a NotebookManager reference, Session, Kernel, or Future - this will store & track it. Hence any request will need exactly 1 lookup ID to find the correct object and operate on it.